src/WingmanDataTrail/SideBar/SideBar.tsx

```diff
@@ -1,146 +1,197 @@
-      model.setState({ [selectedKey]: filters });
-      model.publishEvent(new EventFiltersChanged({ type: type, filters }), true);
-    };
+    const { sections, visibleSection, sectionValues } = model.useState();
 
     return (
       <div className={styles.container}>
-        <div className={styles.topPanel}>
-          <MetricsFilterSection
-            title="Metric prefix filters"
-            items={prefixGroups}
-            selectedValues={selectedMetricPrefixes}
-            onSelectionChange={(filters) => onSelectFilter('prefixes', filters)}
-            loading={loading}
-            dataTestId="metric-prefix-filters"
-          />
-        </div>
-        <div className={styles.bottomPanel}>
-          <MetricsFilterSection
-            title="Metric suffix filters"
-            items={suffixGroups}
-            selectedValues={selectedMetricSuffixes}
-            onSelectionChange={(filters) => onSelectFilter('suffixes', filters)}
-            loading={loading}
-            dataTestId="metric-suffix-filters"
-          />
+        <div className={styles.buttonsBar} data-testid="sidebar-buttons">
+          {sections.map((section) => {
+            const { key, title, icon: iconOrText, disabled, active } = section.state;
+            const visible = visibleSection?.state.key === key;
+            const tooltip = sectionValues.get(key)?.length ? `${title}: ${sectionValues.get(key)?.join(', ')}` : title;
+
+            return (
+              <div
+                key={key}
+                className={cx(styles.buttonContainer, visible && 'visible', active && 'active', disabled && 'disabled')}
+              >
+                <SideBarButton
+                  ariaLabel={title}
+                  disabled={disabled}
+                  visible={visible}
+                  active={active}
+                  tooltip={tooltip}
+                  onClick={() => model.setActiveSection(key)}
+                  iconOrText={iconOrText}
+                />
+              </div>
+            );
+          })}
         </div>
+        {visibleSection && (
+          <div className={styles.content} data-testid="sidebar-content">
+            <IconButton
+              className={styles.closeButton}
+              name="times"
+              aria-label="Close"
+              tooltip="Close"
+              tooltipPlacement="top"
+              onClick={() => model.setActiveSection('')}
+            />
+            {/* :man_shrug: */}
+            {visibleSection instanceof MetricsFilterSection && <visibleSection.Component model={visibleSection} />}
+            {visibleSection instanceof LabelsBrowser && <visibleSection.Component model={visibleSection} />}
+            {visibleSection instanceof BookmarksList && <visibleSection.Component model={visibleSection} />}
+            {visibleSection instanceof Settings && <visibleSection.Component model={visibleSection} />}
```

Comment:
To avoid this, you could use a type guard. For example, this `hasComponent` type guard could go somewhere:

```ts
function hasComponent<T extends SideBarSection>(obj: unknown): obj is T {
  return obj != null && typeof obj === 'object' && 'Component' in obj;
}
```

and then we can clean up the JSX:

```suggestion
            {hasComponent<SideBarSection>(visibleSection) && <visibleSection.Component model={visibleSection} />}
```

Reply:
Thank you for the suggestion Nick! But I'm afraid it does not work. Maybe the way to go is to make all the sections extend from a common ancestor? Happy to pair with you to improve it.